env/io_posix.h

```diff
@@ -49,6 +52,54 @@ class PosixHelper {
                                                size_t* size);
 };
 
+#if defined(ROCKSDB_IOURING_PRESENT)
+struct Posix_IOHandle {
+  struct iovec iov;
+  struct io_uring* iu;
+  std::function<void(const FSReadRequest&, void*)> cb;
+  void* cb_arg;
+  uint64_t offset;
+  size_t len;
+  char* scratch;
+  bool is_finished = false;
+};
+
+inline void UpdateResult(struct io_uring_cqe* cqe, const std::string& file_name,
+                         size_t len, size_t iov_len, bool async_read,
+                         size_t& finished_len, FSReadRequest* req) {
+  if (cqe->res < 0) {
+    req->result = Slice(req->scratch, 0);
+    req->status = IOError("Req failed", file_name, cqe->res);
+  } else {
+    size_t bytes_read = static_cast<size_t>(cqe->res);
+    TEST_SYNC_POINT_CALLBACK("UpdateResults::io_uring_result", &bytes_read);
+    if (bytes_read == iov_len) {
+      req->result = Slice(req->scratch, req->len);
+      req->status = IOStatus::OK();
+    } else if (bytes_read == 0) {
+      if (async_read) {
+        // No  bytes read.
+        req->result = Slice(req->scratch, 0);
+        req->status = IOStatus::IOError("No bytes read");
+      }
+    } else if (bytes_read < iov_len) {
+      assert(bytes_read > 0);
+      if (async_read) {
```

Comment:
Why is async_read case different here? If bytes_read < iov_len, shouldn't we attempt to read the rest like MultiRead does? Unless its the caller's responsibility to call Read/ReadAsync for the remaining bytes.

Reply:
@anand1976 The check is in Poll API in FileSystem. The Read API is in FSRandomAccessFile. Is there a way to call Read API from Poll? Currently the PrefetchBuffer will issue the Read API  if requested bytes are not in the buffer.